File: adventjs/2025/14/solution.ts
type Gift = string | number | boolean
type Workshop = Record<string, any>
type Path = string[]

function findGiftPath(workshop: Workshop, gift: Gift): Path {
  function search(obj: any, target: Gift, path: string[]): string[] | null {
    // Check if current object is the target
    if (obj === target) {
      return path;
    }
    
    // If obj is not an object or is null, skip
    if (typeof obj !== 'object' || obj === null) {
      return null;
    }
    
    // Search through all keys
    for (const key in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        const newPath = [...path, key];
        const result = search(obj[key], target, newPath);
        if (result !== null) {
          return result;
        }
      }
    }
    
    return null;
  }
  
  const result = search(workshop, gift, []);
  return result || [];
}

File: adventjs/2025/14/solution.py
def find_gift_path(workshop: dict, gift: str | int | bool) -> list[str]:
  def search(obj, target, path):
    # Check if current object is the target
    if obj == target:
      return path
    
    # If obj is not a dict, skip
    if not isinstance(obj, dict):
      return None
    
    # Search through all keys
    for key in obj:
      new_path = path + [key]
      result = search(obj[key], target, new_path)
      if result is not None:
        return result
    
    return None
  
  result = search(workshop, gift, [])
  return result if result is not None else []
